ui/vr_webgl.js

```diff
@@ -0,0 +1,611 @@
+      this.eventManager_.listen(this.video_, 'seeked', () => {
+        this.renderGL_();
+      });
+
+      this.eventManager_.listen(document, 'visibilitychange', () => {
+        this.renderGL_();
+      });
+    });
+  }
+
+  /**
+   * @private
+   */
+  initMatrices_() {
+    shaka.ui.Matrix4x4.lookAt(
+        this.viewMatrix_, [0, 0, 0], [1, 0, 0], [0, 1, 0]);
+    shaka.ui.Matrix4x4.getRotation(
+        this.originalQuaternion_, this.viewMatrix_);
+    shaka.ui.Matrix4x4.scale(
+        this.identityMatrix_, this.identityMatrix_, [4.0, 4.0, 4.0]);
+  }
+
+  /**
+   * @private
+   */
+  initGL_() {
+    this.updateViewPort_();
+    this.gl_.viewport(
+        0, 0, this.gl_.drawingBufferWidth, this.gl_.drawingBufferHeight);
+    this.gl_.clearColor(0.0, 0.0, 0.0, 1.0);
+    this.gl_.enable(this.gl_.CULL_FACE);
+    this.gl_.cullFace(this.gl_.FRONT);
+    // Clear the context with the newly set color. This is
+    // the function call that actually does the drawing.
+    this.gl_.clear(this.gl_.COLOR_BUFFER_BIT);
+  }
+
+  /**
+   * @private
+   */
+  initGLShaders_() {
+    const vertexShader = this.getGLShader_(this.gl_.VERTEX_SHADER);
+    const fragmentShader = this.getGLShader_(this.gl_.FRAGMENT_SHADER);
+
+    // Create program
+    this.shaderProgram_ = this.gl_.createProgram();
+    this.gl_.attachShader(this.shaderProgram_, vertexShader);
+    this.gl_.attachShader(this.shaderProgram_, fragmentShader);
+    this.gl_.linkProgram(this.shaderProgram_);
+
+    // If creating the shader program failed, alert
+    if (!this.gl_.getProgramParameter(
+        this.shaderProgram_, this.gl_.LINK_STATUS)) {
+      shaka.log.error('Unable to initialize the shader program: ',
+          this.gl_.getProgramInfoLog(this.shaderProgram_));
+    }
+
+    // Bind data
+    this.vertexPositionAttribute_ = this.gl_.getAttribLocation(
+        this.shaderProgram_, 'a_vPosition');
+    this.gl_.enableVertexAttribArray(this.vertexPositionAttribute_);
+    this.textureCoordAttribute_ = this.gl_.getAttribLocation(
+        this.shaderProgram_, 'a_TexCoordinate');
+    this.gl_.enableVertexAttribArray(this.textureCoordAttribute_);
+  }
+
+  /**
+   * Read and generate WebGL shader
+   *
+   * @param {number} glType Type of shader requested.
+   * @return {?WebGLShader}
+   * @private
+   */
+  getGLShader_(glType) {
+    let source;
+
+    switch (glType) {
+      case this.gl_.VERTEX_SHADER:
+        source = shaka.ui.VRUtils.VERTEX_SPHERE_SHADER;
+        break;
+      case this.gl_.FRAGMENT_SHADER:
+        source = shaka.ui.VRUtils.FRAGMENT_SPHERE_SHADER;
+        break;
+      default:
+        return null;
+    }
+
+    const shader = this.gl_.createShader(glType);
+
+    this.gl_.shaderSource(shader, source);
+
+    this.gl_.compileShader(shader);
+
+    if (!this.gl_.getShaderParameter(shader, this.gl_.COMPILE_STATUS)) {
+      shaka.log.warning(
+          'Error in ' + glType + ' shader' + this.gl_.getShaderInfoLog(shader));
+      return null;
```

Comment:
Also, since the shader programs are constants that the user can't touch, and since it doesn't look like we actually stop setting up VR if we return null here, these should be asserts instead.

Reply:
Done!